src/storage/src/source/postgres/snapshot.rs

```diff
@@ -250,12 +250,29 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
                 .config(&*context.secrets_reader)
                 .await?
                 .tcp_timeouts(config.params.pg_source_tcp_timeouts.clone());
+
             let task_name = format!("timely-{worker_id} PG snapshotter");
 
-            let client = if is_snapshot_leader {
-                let client = connection_config.connect_replication(
+            let mut client = connection_config.connect(
+                &task_name,
+                &context.ssh_tunnel_manager,
+            ).await?;
+
+            // If publication does not exist, we do not need to snapshot any
+            // tables because the replication dataflow will also return.
+            // Exiting here ensures that we don't encounter the missing
+            // publication as a transient error when it should be definite.
```

Comment:
This comment implies there is some place below that errors out when the publication is not there but manifests as a transient error. This sounds like a bug of the code below that we should fix instead of having a check that ensures the buggy code is never run.

The only place I see using the publication is the `mz_postgres_util::publication_info` call. This operator should check if the error from that call is "publication does not exist" and act on it. Also, this operator should *itself* produce the definite errors in that case. It's ok if we double produce them, it makes it easier to reason about each operator separately.

Reply:
Sure thing––I thought the equivalence between `PostgresError` and `DefiniteError` was a little kludgy which was why I chose to use the `DefiniteError`-returning function. This approach wfm, though. And thanks for the pointer on generating the definite errors in both dataflows––hadn't considered that.